source/cppfs/source/posix/LocalFileWatcher.cpp

```diff
@@ -0,0 +1,128 @@
+
+#include <cppfs/posix/LocalFileWatcher.h>
+
+#include <unistd.h>
+#include <limits.h>
+#include <sys/inotify.h>
+
+#include <cppfs/cppfs.h>
+#include <cppfs/FilePath.h>
+#include <cppfs/FileHandle.h>
+#include <cppfs/FileIterator.h>
+#include <cppfs/FileWatcher.h>
+#include <cppfs/posix/LocalFileSystem.h>
+#include <cppfs/posix/LocalFileIterator.h>
+
+
+namespace cppfs
+{
+
+
+LocalFileWatcher::LocalFileWatcher(FileWatcher * fileWatcher, std::shared_ptr<LocalFileSystem> fs)
+: AbstractFileWatcherBackend(fileWatcher)
+, m_fs(fs)
+, m_inotify(-1)
+{
+    // Create inotify instance
+    m_inotify = inotify_init();
+}
+
+LocalFileWatcher::~LocalFileWatcher()
+{
+    // Close watch handles
+    for (auto it : m_watchers) {
+        inotify_rm_watch(m_inotify, it.first);
+    }
+
+    // Close inotify instance
+    close(m_inotify);
+}
+
+AbstractFileSystem * LocalFileWatcher::fs() const
+{
+    // Return file system
+    return static_cast<AbstractFileSystem *>(m_fs.get());
+}
+
+void LocalFileWatcher::add(FileHandle & fh, unsigned int events, RecursiveMode recursive)
+{
+    // Get watch mode
+    uint32_t flags = 0;
+    if (events & FileCreated)  flags |= IN_CREATE;
+    if (events & FileRemoved)  flags |= IN_DELETE;
+    if (events & FileModified) flags |= IN_MODIFY;
+
+    // Create watcher
+    int handle = inotify_add_watch(m_inotify, fh.path().c_str(), flags);
+    if (handle < 0) {
+        return;
+    }
+
+    // Watch directories recursively
+    if (fh.isDirectory() && recursive == Recursive) {
+        // List directory entries
+        for (auto it = fh.begin(); it != fh.end(); ++it)
+        {
+            // Check if entry is a directory
+            FileHandle fh2 = fh.open(*it);
+            if (fh2.isDirectory()) {
+                // Watch directory
+                add(fh2, events, recursive);
+            }
+        }
+    }
+
+    // Associate watcher handle with file handle
+    m_watchers[handle].fileHandle = fh;
+    m_watchers[handle].events     = events;
+    m_watchers[handle].recursive  = recursive;
+}
+
+void LocalFileWatcher::watch()
+{
+    // Create buffer for receiving events
+    size_t bufSize = 64 * (sizeof(inotify_event) + NAME_MAX);
```

Comment:
Where is the 64 coming from?
May we want to increase the number in several years?

Reply:
It's arbitrary, and maybe much larger than we need. In reality, I've only seen 2 events at the max coming in with one call. This function will always be called in a loop, so it doesn't really matter how many events can be read in one go, so I don't think we need to increase it later.